qtfred/src/mission/dialogs/CampaignEditorDialogModel.cpp

```diff
@@ -0,0 +1,801 @@
+#include "CampaignEditorDialogModel.h"
+#include "weapon/weapon.h"
+#include "cutscene/cutscenes.h"
+#include "menuui/mainhallmenu.h"
+#include "stats/scoring.h"
+#include "mission/missiongoals.h"
+#include "mission/missionsave.h"
+
+#include <QPlainTextDocumentLayout>
+
+extern int Skip_packfile_search;
+
+extern void parse_init(bool basic);
+extern void parse_mission_info(mission *mn, bool basic);
+extern void parse_events(mission *mn);
+extern void parse_goals(mission* mn);
+extern int Subsys_status_size;
+
+namespace fso {
+namespace fred {
+namespace dialogs {
+
+static QString loadFile(QString file, const QString& campaignType) {
+	if (file.isEmpty()) {
+		mission_campaign_clear();
+		Campaign.type = campaignType.isEmpty() ? CAMPAIGN_TYPE_SINGLE :
+												 CampaignEditorDialogModel::campaignTypes.indexOf(campaignType);
+		return {};
+	}
+	//FRED is to enforce that only on new campaigns a campaign type may be given
+	Assert(campaignType.isEmpty());
+	parse_init(false);
+	if (mission_campaign_load(qPrintable(file.replace('/',DIR_SEPARATOR_CHAR)), nullptr, 0))
+		return {};
+
+	return Campaign.filename;
+}
+
+static void initShips(const SCP_vector<ship_info>::const_iterator &s_it, CheckedDataListModel<std::ptrdiff_t> &model){
+	std::ptrdiff_t shpIdx{ std::distance(Ship_info.cbegin(), s_it) };
+	if (s_it->flags[Ship::Info_Flags::Player_ship]) {
+		model.initRow(
+				s_it->name,
+				shpIdx,
+				static_cast<bool>(Campaign.ships_allowed[static_cast<size_t>(shpIdx)]));
+}
+}
+
+static void initWeps(const SCP_vector<weapon_info>::const_iterator &w_it, CheckedDataListModel<std::ptrdiff_t> &model){
+	std::ptrdiff_t wepIdx{ std::distance(Weapon_info.cbegin(), w_it) };
+	for (const ship_info& shp: Ship_info) {
+		if (shp.flags[Ship::Info_Flags::Player_ship]
+			&& shp.allowed_weapons[static_cast<size_t>(wepIdx)]
+			&& !model.contains(w_it->name)) {
+			model.initRow(
+					w_it->name,
+					wepIdx,
+					static_cast<bool>(Campaign.weapons_allowed[static_cast<size_t>(wepIdx)]));
+}
+}
+}
+
+static SCP_vector<SCP_string> getMissions(){
+	Skip_packfile_search = 1;
+	SCP_vector<SCP_string> missions;
+	cf_get_file_list(missions, CF_TYPE_MISSIONS, "*fs2", CF_SORT_NAME);
+	Skip_packfile_search = 0;
+	return missions;
+}
+
+static bool parseMnPart(mission *mn, const char *filename){
+	try {
+		read_file_text(filename, CF_TYPE_MISSIONS);
+
+		parse_init(false);
+		Subsys_index = 0;
+		Subsys_status_size = 0;
+		if (Subsys_status != nullptr) {
+			vm_free( Subsys_status );
+			Subsys_status = nullptr;
+		}
+
+		parse_mission_info(mn, true);
+		skip_to_start_of_string("#Events");
+		Num_mission_events = 0;
+		parse_events(mn);
+		Num_goals = 0;
+		parse_goals(mn);
+	}  catch (const parse::ParseException& ) {
```

Comment:
Too many spaces between `}` and `catch` ?

There's also lopsided space padding inside of the catch expression's `( )`